pkg/controller/che/che_controller.go

```diff
@@ -574,44 +559,60 @@ func (r *ReconcileChe) Reconcile(request reconcile.Request) (reconcile.Result, e
 	}
 
 	// create Che service and route
-	cheLabels := deploy.GetLabels(instance, util.GetValue(instance.Spec.Server.CheFlavor, deploy.DefaultCheFlavor))
+	serviceStatus := deploy.SyncCheServiceToCluster(instance, clusterAPI)
+	if !tests {
+		if !serviceStatus.Continue {
+			logrus.Info("Waiting on service 'che-host' to be ready")
+			if serviceStatus.Err != nil {
+				logrus.Error(serviceStatus.Err)
+			}
 
-	if _, err := deploy.NewCheService(instance, cheLabels, r); err != nil {
-		return reconcile.Result{}, err
+			return reconcile.Result{Requeue: serviceStatus.Requeue}, serviceStatus.Err
+		}
 	}
+
 	if !isOpenShift {
-		cheIngress := deploy.NewIngress(instance, cheFlavor, "che-host", 8080)
-		if err := r.CreateNewIngress(instance, cheIngress); err != nil {
-			return reconcile.Result{}, err
+		ingressStatus := deploy.SyncIngressToCluster(instance, cheFlavor, "che-host", 8080, clusterAPI)
+		if !tests {
+			if !ingressStatus.Continue {
+				logrus.Info("Waiting on ingress 'che-host' to be ready")
+				if ingressStatus.Err != nil {
+					logrus.Error(ingressStatus.Err)
+				}
+
+				return reconcile.Result{Requeue: ingressStatus.Requeue}, ingressStatus.Err
+			}
 		}
+
 		cheHost := ingressDomain
 		if ingressStrategy == "multi-host" {
 			cheHost = cheFlavor + "-" + instance.Namespace + "." + ingressDomain
 		}
-		if len(instance.Spec.Server.CheHost) == 0 {
+		if instance.Spec.Server.CheHost != cheHost {
```

Comment:
The logic has changed here, right ? Do we really want to always override the CR `CheHost` value, while previously we used to fill it only when  empty ?

Reply:
I am not sure if we did it correctly before that. 
We should keep it up to date.

Reply:
I see. Thanks for the explanation